### PR TITLE
Add InitiateMaintenance New Command 

### PR DESCRIPTION
**Pull request by vanbasten2323**
## Description

1. Added new Initiate-Maintenance command to perform maintenance on VMs.
2. Added MaintenanceStatus new field in the response of Get-AzureVM command.
---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@azuresdkci add to whitelist

File: src/Common/Commands.Common.Authentication/Commands.Common.Authentication.csproj [hunk 1]
```diff
@@ -95,6 +95,10 @@
       <HintPath>..\..\packages\Microsoft.Bcl.Async.1.0.168\lib\net40\Microsoft.Threading.Tasks.Extensions.Desktop.dll</HintPath>
       <Private>True</Private>
     </Reference>
+    <Reference Include="Microsoft.WindowsAzure.Management.Compute, Version=14.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (shahabhijeet) [inline src/Common/Commands.Common.Authentication/Commands.Common.Authentication.csproj]: @vanbasten2323 any reason Authentication need to take dependency on Compute?

File: src/Common/Commands.Common.Storage/Commands.Common.Storage.csproj [hunk 2]
```diff
@@ -111,6 +111,10 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\packages\Microsoft.WindowsAzure.Management.4.1.1\lib\net40\Microsoft.WindowsAzure.Management.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.WindowsAzure.Management.Compute, Version=14.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (shahabhijeet) [inline src/Common/Commands.Common.Storage/Commands.Common.Storage.csproj]: @vanbasten2323 similarly does storage need to take dependency on Compute?

File: src/Common/Commands.Common/Commands.Common.csproj [hunk 3]
```diff
@@ -97,6 +97,10 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\packages\Microsoft.WindowsAzure.Management.4.1.1\lib\net40\Microsoft.WindowsAzure.Management.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.WindowsAzure.Management.Compute, Version=14.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (shahabhijeet) [inline src/Common/Commands.Common/Commands.Common.csproj]: Same as above, we should not take this cross dependencies

**Comment by cormacpayne** [comment]
@vanbasten2323 please make sure you have pulled the latest changes form the **dev** branch - it appears there are changes in your PR that have already been merged (_e.g.,_ changing `result` to `resultList` in a few files was a fix that @hyonholee had in a PR that was merged)

File: src/ServiceManagement/Common/Commands.ServiceManagement.Common/packages.config [hunk 4]
```diff
@@ -14,5 +14,6 @@
   <package id="Microsoft.Rest.ClientRuntime" version="2.3.5" targetFramework="net45" />
   <package id="Microsoft.Rest.ClientRuntime.Azure.Authentication" version="2.2.9-preview" targetFramework="net45" />
   <package id="Microsoft.WindowsAzure.Management" version="4.1.1" targetFramework="net45" />
+  <package id="Microsoft.WindowsAzure.Management.Compute" version="14.0.0" targetFramework="net45" />
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ServiceManagement.Common/packages.config]: Remove this.  This project does not need Microsoft.WindowsAzure.Management.Compute

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs [hunk 5]
```diff
@@ -48,13 +55,17 @@ public string Name
...
+            ValueFromPipeline = true,
             HelpMessage = "The Virtual Machine to restart.",
             ParameterSetName = RestartInputParameterSet)]
         [Parameter(Mandatory = true,
-            ValueFromPipelineByPropertyName = true,
+            ValueFromPipeline = true,
             HelpMessage = "The Virtual Machine to redeploy.",
             ParameterSetName = RedployInputParameterSet)]
+        [Parameter(Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
```
Comment (AzureRT) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs]: Change it to ValueFromPipeline = true, as the other parameter set

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 [hunk 6]
```diff
@@ -1007,3 +1007,61 @@ function Run-RedeployVirtualMachineTest
...
+	$userName = "pstestuser";
+	$password = $PLACEHOLDER;
+
+	# Test
+	New-AzureService -ServiceName $svcName -Location $location;
+
+	try
+	{
+		New-AzureQuickVM -Windows -ImageName $imgName -Name $vmName -ServiceName $svcName -AdminUsername $userName -Password $password;
+		Start-Sleep -s 300;
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Commented out this line for Playback mode such as below:

#Start-Sleep -s 300; #Uncomment this line for record mode testing.

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 [hunk 7]
```diff
@@ -1007,3 +1007,61 @@ function Run-RedeployVirtualMachineTest
...
+
+		# Test Initiate Maintenance
+		$result = Restart-AzureVM -InitiateMaintenance -ServiceName $svcName -Name $vmName;
+
+		$vm = Get-AzureVM -ServiceName $svcName -Name $vmName
+		Assert-NotNull $vm.MaintenanceStatus; 
+	}
+	catch
+	{
+        Assert-True {$result.Result.Contains("User initiated maintenance on the Virtual Machine was successfully completed.")};
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Align space

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 [hunk 8]
```diff
@@ -1007,3 +1007,61 @@ function Run-RedeployVirtualMachineTest
...
     }
 }
+
+# Test Initiate Maintenance VM
+function Run-InitiateMaintenanceTest
+{
+	# Depending on the environment, the initiate maintenance operation may return 200
+	# or 400 with error message like "User initiated maintenance on the virtual machine was
+	# successfully completed". Both are expected reponse.
+    # To continue script, $ErrorActionPreference should be set to 'SilentlyContinue'.
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Align space

File: src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1 [hunk 9]
```diff
@@ -1007,3 +1007,61 @@ function Run-RedeployVirtualMachineTest
...
+	$tempErrorActionPreference = $ErrorActionPreference;
+    $ErrorActionPreference='SilentlyContinue';
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Resources/ServiceManagement/ServiceManagementTests.ps1]: Align space

File: src/ServiceManagement/Common/Commands.ScenarioTest/Commands.ScenarioTest.csproj [hunk 10]
```diff
@@ -275,7 +275,7 @@
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </None>
     <None Include="SessionRecords\Microsoft.WindowsAzure.Commands.ScenarioTest.ChefExtensionTests\TestSetAzureVMChefExtensionDaemonTask.json">
-	<CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    <CopyToOutputDirectory>Always</CopyToOutputDirectory>
```
Comment (AzureRT) [inline src/ServiceManagement/Common/Commands.ScenarioTest/Commands.ScenarioTest.csproj]: Align space

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs [hunk 11]
```diff
@@ -48,13 +55,17 @@ public string Name
...
+            HelpMessage = "The Virtual Machine to initiate maintenance.",
+            ParameterSetName = InitiateMaintenanceInputParameterSet)]
```
Comment (shahabhijeet) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs]: @vanbasten2323 why VM parameter is not part of InitiateMaintenanceByNameParameterSet?

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs [hunk 12]
```diff
@@ -48,13 +55,17 @@ public string Name
         }
 
         [Parameter(Mandatory = true,
-            ValueFromPipelineByPropertyName = true,
+            ValueFromPipeline = true,
```
Comment (vanbasten2323) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs]: Yes. The test has already covered that.

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs (hunk 11 above)
Comment (vanbasten2323) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/PersistentVMs/RestartAzureVM.cs]: This command has two parameter sets: InitiateMaintenanceByNameParameterSet which takes a VM name and InitiateMaintenanceInputParameterSet which takes a VM parameter.

**Comment by hyonholee** [comment]
LGTM.  Thanks.

**Comment by shahabhijeet** [comment]
please add change log for your changes and describe them.

**Pull request merged**